### PR TITLE
[14.0][FIX] sale_advance_payment:  Fix residual amount not being zero on fully invoice

### PR DESCRIPTION
**Pull request by JasminSForgeFlow**

### CODE REVIEW

File: sale_advance_payment/models/sale.py [hunk 1]
```diff
@@ -61,7 +62,14 @@ def _compute_advance_payment(self):
             advance_amount = 0.0
             for line in mls:
                 line_currency = line.currency_id or line.company_id.currency_id
-                line_amount = line.amount_currency if line.currency_id else line.balance
+                # Exclude reconciled pre-payments amount because once reconciled
+                # the pre-payment will reduce bill residual amount like any
```
Comment (LoisRForgeFlow) [inline sale_advance_payment/models/sale.py]: bill -> invoice

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 2]
```diff
@@ -183,7 +181,7 @@ def test_sale_advance_payment(self):
 
         self.assertEqual(self.sale_order_1.amount_residual, 3280)
 
-        # Confirm Sale Order
+        # Confirm Purchase Order
```
Comment (LoisRForgeFlow) [inline sale_advance_payment/tests/test_sale_advance_payment.py]: Sales Order

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 3]
```diff
@@ -156,7 +156,6 @@ def test_sale_advance_payment(self):
             .create(
                 {
                     "journal_id": self.journal_eur_bank.id,
-                    "payment_type": "inbound",
```
Comment (LoisRForgeFlow) [inline sale_advance_payment/tests/test_sale_advance_payment.py]: SO payments aren't inbound payments? why do you remove this?

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 4]
```diff
@@ -218,56 +214,147 @@ def test_sale_advance_payment(self):
...
-        # Create Invoice
-        invoice = self.sale_order_1._create_invoices()
-        invoice.action_post()
-
-        # Compare payments
-        rate = self.currency_rate.rate
-        payment_list = [100 * rate, 200, 250 * rate, 400]
-        payments = json.loads(invoice.invoice_outstanding_credits_debits_widget)
-        result = [d["amount"] for d in payments["content"]]
-        self.assertEqual(set(payment_list), set(result))
```
Comment (LoisRForgeFlow) [inline sale_advance_payment/tests/test_sale_advance_payment.py]: Why do you remove this part? This was a valid test

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 5]
```diff
@@ -218,56 +214,147 @@ def test_sale_advance_payment(self):
...
-
-    def test_sale_advance_payment_outgoing(self):
+    def test_02_residual_amount_with_bill(self):
```
Comment (LoisRForgeFlow) [inline sale_advance_payment/tests/test_sale_advance_payment.py]: bill -> invoice

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 6]
```diff
@@ -218,56 +214,147 @@ def test_sale_advance_payment(self):
...
+        ).create(
+            {
+                "amount": 1200.0,
+                "group_payment": True,
+                "payment_difference_handling": "open",
+            }
+        )._create_payments()
+        self.assertEqual(self.sale_order_1.amount_residual, 2200)
+
+        # Reconciling the pre-payment should not affect amount_residual in PO.
```
Comment (LoisRForgeFlow) [inline sale_advance_payment/tests/test_sale_advance_payment.py]: SO

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 7]
```diff
@@ -218,56 +214,147 @@ def test_sale_advance_payment(self):
...
+        (
+            liquidity_lines,
+            counterpart_lines,
+            writeoff_lines,
+        ) = pre_payment._seek_for_lines()
+        (
+            counterpart_lines
+            + invoice.line_ids.filtered(
+                lambda line: line.account_internal_type == "payable"
```
Comment (LoisRForgeFlow) [inline sale_advance_payment/tests/test_sale_advance_payment.py]: receivable

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 8]
```diff
@@ -218,56 +214,147 @@ def test_sale_advance_payment(self):
...
+        # Partially reconciling the pre-payment should not affect amount_residual in PO.
+        (
+            liquidity_lines,
+            counterpart_lines,
+            writeoff_lines,
+        ) = pre_payment._seek_for_lines()
+        (
+            counterpart_lines
+            + invoice.line_ids.filtered(
+                lambda line: line.account_internal_type == "payable"
```
Comment (LoisRForgeFlow) [inline sale_advance_payment/tests/test_sale_advance_payment.py]: receivable.

you are probably not reconciling anything here, can you add an assert for the amount_residual on the invoice after this?

`self.assertEqual(invoice.amount_residual, 0)`

**Review by LoisRForgeFlow — CHANGES_REQUESTED** [review]
Some comments

**Comment by LoisRForgeFlow** [comment]
For the record, this PR is aligning behavior with purchase_advance_payment (see https://github.com/OCA/purchase-workflow/commit/82d73cb54fb22b873cf4493406184d023b31f944 and https://github.com/OCA/purchase-workflow/pull/1324)

**Review by LoisRForgeFlow — APPROVED** [review]
LGTM now :+1:

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 9]
```diff
@@ -21,13 +22,13 @@ def setUpClass(cls):
 
         # Products
         cls.product_1 = cls.env["product.product"].create(
-            {"name": "Desk Combination", "invoice_policy": "order"}
+            {"name": "Desk Combination", "type": "product", "invoice_policy": "order"}
```
Comment (MateuGForgeFlow) [inline sale_advance_payment/tests/test_sale_advance_payment.py]: I would remove the "type": "product", from the products. This crashes the tests if you have a clean database with only the sale_advance_payment (as the only dependency is sale).
We did this modification in another PR recently. Perhaps the code is not updated with 14.0. Maybe there are some other conflicts. https://github.com/OCA/sale-workflow/pull/1847

Comment (JasminSForgeFlow) [reply sale_advance_payment/tests/test_sale_advance_payment.py]: Hi @MateuGForgeFlow,

Update change as you suggest, Please check,

Thanks

Comment (MateuGForgeFlow) [reply sale_advance_payment/tests/test_sale_advance_payment.py]: Hi @JasminSForgeFlow! There's still the "type": "service" for product 3. Thank you!

Comment (JasminSForgeFlow) [reply sale_advance_payment/tests/test_sale_advance_payment.py]: Hi @MateuGForgeFlow ,
Update changes for 3rd product.
Please check.
Thanks

**Review by MateuGForgeFlow — APPROVED** [review]

File: sale_advance_payment/tests/test_sale_advance_payment.py (hunk 9 above)
Comment (MateuGForgeFlow) [reply sale_advance_payment/tests/test_sale_advance_payment.py]: Thank you! @JasminSForgeFlow

File: sale_advance_payment/tests/test_sale_advance_payment.py [hunk 10]
```diff
@@ -117,7 +118,7 @@ def setUpClass(cls):
             }
         )
 
-    def test_sale_advance_payment(self):
+    def test_01_sale_advance_payment(self):
```
Comment (rousseldenis) [inline sale_advance_payment/tests/test_sale_advance_payment.py:121]: FYI numbering function does not ensure execution order.

Comment (LoisRForgeFlow) [reply sale_advance_payment/tests/test_sale_advance_payment.py:121]: I thought it did, the test aren't executed in alphabetical order?

**Comment by LoisRForgeFlow** [comment]
@rousseldenis good to merge?

**Pull request merged**